models/repo_collaboration.go

```diff
@@ -210,3 +210,21 @@ func (repo *Repository) getRepoTeams(e Engine) (teams []*Team, err error) {
 func (repo *Repository) GetRepoTeams() ([]*Team, error) {
 	return repo.getRepoTeams(x)
 }
+
+// IsOwnerMemberCollaborator checks if a provided user is the owner, a collaborator or a member of a team in a repository
+func (repo *Repository) IsOwnerMemberCollaborator(userID int64) (bool, error) {
+	if repo.OwnerID == userID {
```

Comment:
Why not `models.accessLevel()`? (site admin should perhaps accounted for here too).

Reply:
Which access level and unit we should test? It can't be any of the Read units because then any public project would assign everyone as a member, Write would likely be similarly to high. I also don't think the site admin should be automatically considered a member of a project.

Reply:
Sorry, you're right, `models.accessLevel()` is not good enough because it doesn't query by unit type. But the `access` table already contains who has access to what and we can query by `UnitTypeCode` (and again, perhaps also site admin). I don't think we should check commits from people that doesn't have access to code (e.g. teams that have only issue or wiki access).

Note that the `access` table only contains explicit permissions (as from ownership, collaboration or team membership).

Reply:
```go

// IsOwnerMemberCollaborator checks if a provided user is the owner, a collaborator or a member of a team in a repository
func (repo *Repository) IsOwnerMemberCollaborator(userID int64) (bool, error) {
	if repo.OwnerID == userID {
		return true, nil
	}

	sess := x.NewSession()
	if err := sess.Begin(); err != nil {
		return false, err
	}
	defer sess.Close()

	if access, err := hasAccess(sess, userID, repo); !access || err != nil {
		return access, err
	}

	user, err := getUserByID(sess, userID)
	if err != nil {
		return false, err
	}

	perm, err := getUserRepoPermission(sess, repo, user)
	if err != nil {
		return false, err
	}

	return perm.CanRead(UnitTypeCode), sess.Commit()
}
```

Reply:
damn `hasAccess` is not correct.

Reply:
Oh, my, I need to learn to bite my tongue 🤐:

>
> But the access table already contains who has access to what and we can query by `UnitTypeCode`.

That would be my `user_repo_unit` table, _which doesn't exist yet_. The `access` table does not discriminate by `UnitType` and that's the essence of the problem with that table.

So for this to work, the honest-to-goodness thing to do would be using `GetUserRepoPermission()` and then `p.CanRead()` or `p.CanWrite()`, but that would be way too slow.

So, I'd suggest we go back to your approach (current `IsOwnerMemberCollaborator()`) but with one addition: I'd make sure the team has `UnitTypeCode` access by joining with `team_unit`.

I'm truly sorry I've got you going in circles. 😔

Reply:
OK I've done that.